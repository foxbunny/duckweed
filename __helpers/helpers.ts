/**
 * (c) 2017 Hajime Yamasaki Vukelic
 * All rights reserved.
 */

const pause = (delay: number = 0): Promise<void> => {
  return new Promise((res) => {
    setTimeout(res, delay);
  });
};

export {
  pause,
};
