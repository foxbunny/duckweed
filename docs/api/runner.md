# `duckweed.runner()`

```javascript
(model, actions, view, options) => undefined
```

The runner function kick-starts a Duckweed application. It takes the following
arguments.

 param       | type            | description
-------------|-----------------|---------------------------------------
 model       | any type        | The initial application state
 actions     | object          | Mapping from key to message handler
 view        | function        | View function
 options     | object          | Runner options

## Model

Model can be any value or even `undefined` and `null`. Note that this object is
generally not mutated but replaced when actions call a [`patch()`](./patch.md)
function, so it only represents the initial state of the application.

## Actions

The actions object represents a mapping between action address (the first
element in a message), and the message handler. The message handler function has
the following signature:

```javascript
(patch, [...data]) => undefined
```

The message handler function can be either sync or async. Duckweed will not use
its return value and it, therefore, does not care if we use an async function.

The first argument is always the [`patch()`](./patch.md) function which is used
to *patch* the model. This function can be called multiple times within the
message handler, so we can use it to set intermediate states during asyncrhonous
operations.

Here's an example:

```javascript
const actions = {
  getData: async (patch, recordId) => {
    patch((model) => ({
      ...model,
      loading: true,
    }));
    const response = await xhr.get("/data/" + recordId);
    patch((model) => ({
      ...model,
      loading: false,
      data: response.json,
    }));
  },
};
```

When the message handler is handling DOM events or Snabbdom hooks, it may
receive additional arguments (event object, `VNode` objects, etc) if we are
using the [`duckweed.html()`](./html.md) factory to generate our VDOM.

## View

A view function that is passed to runner will have the following signature:

```javascript
({model, act}) => VNode
```

The props object is passed to the function with two keys, `model` and `act`. The
`model` is the current application state. The [`act()`](./act.md) function is
the action handler (yes, bad name), which can be used to transmit messages.

## Options

The runner takes the following options:

interface RunnerOptions {
  root?: string | Element | VNode;
  patch?: VDOMPatchFunction;
  plugins?: Plugin[];
  middleware?: PatchMiddleware[];
}

### `root` (string, `Element`, or `VNode`)

The root element is the first argument that's usually passed to Snabbdom's
[`patch()`](https://github.com/snabbdom/snabbdom#patch). If you are using the
default `patch` or a `patch` function created by using the Snabbdom API, this
should be a CSS3 selector, an `Element` object, or a Snabbdom `VNode` object.

By passing a `VNode` object, we can continue the rendering from some previous
state created by this or some other app (provided that the DOM state matches the
`VNode` object).

**default:** `"#app"`

### `patch` (function)

The patch option can be used to specify an alternative `patch` function. This is
usually necessary if we want to use a different set of Snabbdom modules, or even
completely modify the Duckweed behavior by using our own patch function. The
details of writing a custom `patch()` implementation is beyond the scope of this
documentation.

The default value comes from the `duckweed/html` module, and uses the following
Snabbdom and Duckweed modules:

- snabbdom/modules/class
- snabbdom/modules/eventlisteners
- snabbdom/modules/props
- snabbdom/modules/style
- duckweed/modules/documentevents
- duckweed/modules/keyevents
- duckweed/modules/routeevents

See the [Non-standard events](../guide/non-standard-events) guide for more
information on what the Duckweed modules do.

**default:** `duckweed/html#patch`

### `plugins` (array of plugin objects)

Plugin objects have the following structure:

```javascript
{
  actions: object,
  init(act) => undefined
}
```

The `actions` object has the same strucutre as `actions` passed to the runner.
The init function takes a message handler and is invoked once right before the
first render.

See the [Writing plugins](../guide/plugins.md) for more information on how to
write plugins.

**default:** `[]`

### `middleware` (array of middleware functions)

Middleware functions have the following signature:

```javascript
(patchFunction) => (model) => model
```

The middlware functions take a patch function, which receives a model and
returns a patched model, and returns a patch function with modified behavior.

The middleware functions will be piped (first to last) such that when
[`patch()`](./patch.md) is invoked in a message handler, the call goes from the
first middleware to last, then through the actual patch callback, and back up to
the first middleware.

See [Writing middlware](../guide/middlware.md) guide for more information.

**default:** `[]`

[Documentation index](../main.md) | [API reference index](./main.md)
