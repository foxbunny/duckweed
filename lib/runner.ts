/**
 * (c) 2017 Hajime Yamasaki Vukelic
 * All rights reserved.
 */

import {VNode} from "snabbdom/vnode";

import html, {patch} from "./html";
import * as is from "./is";

type PatchFunction<T = any> = (model: T) => T;

type ModelPatcher<T = any> = (fn: PatchFunction<T>) => void;

interface ActionHandler {
  (...message: any[]): (...eventArgs: any[]) => any;
  as(...message: any[]): ActionHandler;
}

interface Props {
  model: any;
  act: ActionHandler;
}

type ViewFunction = (props: Props) => VNode;

type RenderFunction = (handler: ActionHandler) => void;

interface Actions<T = any> {
  [action: string]: (patch: ModelPatcher<T>, ...args: any[]) => Promise<void>;
}

interface RunnerState<T = any> {
  vnodes: Element | VNode;
  model: T;
  nextRenderId: null | number;
}

interface RunnerOptions {
  root?: string | Element | VNode;
}

/**
 * Clears the timer if one was set by the patch function.
 */
const cancelNextRender = (state: RunnerState): void => {
  if (state.nextRenderId) {
    clearTimeout(state.nextRenderId);
    state.nextRenderId = null;
  }
};

/**
 * Cancel the next-scheduled render, and reschedule another render
 */
const setNextRender = (state: RunnerState, render: RenderFunction): void => {
  cancelNextRender(state);
  state.nextRenderId = setTimeout(render);
};

/**
 * Create a renderer function
 *
 * The renderer function will keep updating the vnodes stored in the runner
 * state using a specified view function.
 */
const createRenderer = (state: RunnerState, view: ViewFunction) => {
  return (actionHandler: ActionHandler) => {
    const props: Props = {model: state.model, act: actionHandler};
    state.vnodes = patch(state.vnodes, html(view, props));
    state.nextRenderId = null;
  };
};

/**
 * Construct the final argument list for action handlers
 *
 * This function takes the user arguments (which are supplied in the props), and
 * event handler callback arguments (usually an `Event` object for event
 * handlers, `VNode` objects for hooks, or some special object for synthetic
 * events), and combines those two as needed.
 *
 * This method generates various convenience arguments, like input value for
 * input events, checkbox state for checkbox toggle events, and so on.
 */
const actionArgs = (userArgs: any[], eventCallbackArgs: any[]) => {
  const first = eventCallbackArgs[0];
  if (is.vnode(first)) {
    // This is mostly for hooks. We add the vnode objects to args.
    return userArgs.concat(eventCallbackArgs);
  } else if (is.changeEvent(first) && is.checkbox(first.target)) {
    return userArgs.concat(first.target.checked, first.target.value);
  } else if (is.inputEvent(first) && is.input(first.target)) {
    // For convenience, process events and extract implied arguments
    first.preventDefault();
    return userArgs.concat(first.target.value);
  } else if (is.pathData(first)) {
    return userArgs.concat(first);
  }
  return userArgs;
};

const actionHandlerFactory = (patcher: ModelPatcher, actions: Actions, baseArgs: any[] = []): ActionHandler => {
  const handler = (action: any, ...args: any[]) => (...eventArgs: any[]) => {
    if (action == null) {
      return;
    }
    const actionFn = actions[action];
    actionFn(patcher, ...actionArgs(baseArgs.concat(args), eventArgs));
  };
  (handler as any).as = (...args: any[]) => actionHandlerFactory(patcher, actions, args);
  return handler as ActionHandler;
};

/**
 * Create an action handler
 *
 * Action handler is a proxy event/hook handler factory which allows the user to
 * specify messages which will then be tied to action handlers when the events
 * trigger.
 *
 * A message constists of an action identifier, and zero or more arbitrary
 * user-specified arguments. The message is specified in the prop, and it is
 * passsed to the action handler, which returns an event handler that is used by
 * Snabbdom to handle the events. When an event is triggered, the control is
 * returned to the action handler which uses the original message to determine
 * which action handler will be invoked.
 */
const createActionHandler = <T = any>(
  state: RunnerState,
  actions: Actions<T>,
  render: RenderFunction,
): ActionHandler => {
  const patcher = (fn: PatchFunction<T>) => {
    state.model = fn(state.model);
    setNextRender(state, () => render(handler));
  };
  const handler = actionHandlerFactory(patcher, actions);
  return handler;
};

const DEFAULT_OPTIONS = {
  root: "#app",
};

/**
 * Create and start a new application runtime
 *
 * The runner function takes a model, actions mapping, view function, and an
 * optional root element selector (defaults to "#app"). It then kicks off the
 * render process, rendering the initial view onto the root element (root
 * element is replaced in the process).
 */
const runner = <T = any> (model: T, actions: Actions<T>, view: ViewFunction, options: RunnerOptions = {}) => {
  const opt = {
    ...DEFAULT_OPTIONS,
    ...options,
  };

  const state: RunnerState<T> = {
    model,
    nextRenderId: null,
    vnodes: is.str(opt.root) ? (document.querySelector(opt.root) as Element) : opt.root,
  };

  // Prepare the engine
  const render = createRenderer(state, view);
  const actionHandler = createActionHandler<T>(state, actions, render);

  // Start rendering
  render(actionHandler);
};

export {
  PatchFunction,
  ModelPatcher,
  ActionHandler,
  Actions,
  Props,
  ViewFunction,
  RunnerState,
  RunnerOptions,
  runner,
};
export default runner;
