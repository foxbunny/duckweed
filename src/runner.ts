/**
 * (c) 2017 Hajime Yamasaki Vukelic
 * All rights reserved.
 */

import {VNode} from "snabbdom/vnode";

import {patch} from "./html";
import * as is from "./is";

type PatchFunction<T = any> = (model: T) => T;

type Scope = Array<string | number>;

interface ModelPatcher<T = any> {
  (fn: PatchFunction<T>): void;
  as<S = any>(scope: Scope): ModelPatcher<S>;
}

interface Actions<T = any> {
  [action: string]: (patch: ModelPatcher<T>, ...args: any[]) => void | Promise<void>;
}

interface ActionHandler {
  (...message: any[]): (...eventArgs: any[]) => any;
  prefix: any[];
  as(...message: any[]): ActionHandler;
}

interface Props {
  model: any;
  act: ActionHandler;
}

type ViewFunction = (props: Props) => VNode;

type RenderFunction = (handler: ActionHandler) => void;

type PatchMiddleware = (fn: PatchFunction) => PatchFunction;

interface RunnerState<T = any> {
  vnodes: Element | VNode;
  model: T;
  nextRenderId: null | number;
}

interface RunnerOptions {
  root?: string | Element | VNode;
  middleware?: PatchMiddleware[];
}

/**
 * Clears the timer if one was set by the patch function.
 */
const cancelNextRender = (state: RunnerState): void => {
  if (state.nextRenderId) {
    clearTimeout(state.nextRenderId);
    state.nextRenderId = null;
  }
};

/**
 * Cancel the next-scheduled render, and reschedule another render
 */
const setNextRender = (state: RunnerState, render: RenderFunction): void => {
  cancelNextRender(state);
  state.nextRenderId = setTimeout(render);
};

/**
 * Create a renderer function
 *
 * The renderer function will keep updating the vnodes stored in the runner
 * state using a specified view function.
 */
const createRenderer = (state: RunnerState, view: ViewFunction) => {
  return (actionHandler: ActionHandler) => {
    state.vnodes = patch(state.vnodes, view({model: state.model, act: actionHandler} as Props));
    state.nextRenderId = null;
  };
};

/**
 * Retrieves the value within an object, at given scope.
 */
const scopeGet = (scope: Array<string | number>, object: any): any => {
  return scope.length
    ? scopeGet(scope.slice(1), object[scope[0]])
    : object;
};

/**
 * Returns a copy of the object with the value assigned to the property at specified scope
 */
const scopeSet = (scope: Scope, val: any, object: any): any => {
  if (scope.length === 0) {
    return val;
  }
  const [first, ...rest] = scope;
  return Array.isArray(object)
    ? (() => {
      const copy = object.concat([]);
      copy[first as number] = scopeSet(rest, val, copy[first as number]);
      return copy;
    })()
    : {...object, [first]: scopeSet(rest, val, object[first])};
};

const scopePatch = (scope: Scope, fn: (arg: any) => any, object: any): any =>
  scopeSet(scope, fn(scopeGet(scope, object)), object);

const createPatcher = <T = any>(
  state: RunnerState,
  middleware: PatchMiddleware[],
  patchCallback: () => void,
  scope: Scope = [],
  parentScope: Scope = [],
  scopeCallback: (model: any) => any = (model) => model,
): ModelPatcher<T> => {
  const middlewareStack = middleware.reduce((m1, m2) => {
    return (fn: PatchFunction) => m1(m2(fn));
  }, (fn: PatchFunction) => fn);
  const mutate: PatchMiddleware = (fn: PatchFunction<T>) => (model: any) => {
    const updated = scope
      ? scopePatch(scope, fn, model)
      : fn(model);
    return scopePatch(parentScope, scopeCallback, updated);
  };
  const patcher = (fn: PatchFunction<T>) => {
    state.model = middlewareStack(mutate(fn))(state.model);
    patchCallback();
  };
  (patcher as any).as = <S = any>(childScope: Scope, parentCallback: (model: any) => any): ModelPatcher<S> => {
    const patcherScope = scope ? scope.concat(childScope) : childScope;
    return createPatcher(state, middleware, patchCallback, patcherScope, scope, parentCallback);
  };
  return patcher as ModelPatcher<T>;
};

const actionHandlerFactory = (patcher: ModelPatcher, actions: Actions, prefix: any[] = []): ActionHandler => {
  const handler = (...args: any[]) => (...eventArgs: any[]) => {
    const [action, ...actionArgs] = prefix.concat(args, eventArgs);
    if (action == null) {
      return;
    }
    const actionFn = actions[action];
    if (!actionFn) {
      throw Error(`No action found for message [${action}, ${actionArgs.join(", ")}]`);
    }
    actionFn(patcher, ...actionArgs);
  };
  (handler as any).as = (...args: any[]) => {
    return actionHandlerFactory(patcher, actions, prefix.concat(args));
  };
  (handler as any).prefix = prefix;
  return handler as ActionHandler;
};

/**
 * Create an action handler
 *
 * Action handler is a proxy event/hook handler factory which allows the user to
 * specify messages which will then be tied to action handlers when the events
 * trigger.
 *
 * A message constists of an action identifier, and zero or more arbitrary
 * user-specified arguments. The message is specified in the prop, and it is
 * passsed to the action handler, which returns an event handler that is used by
 * Snabbdom to handle the events. When an event is triggered, the control is
 * returned to the action handler which uses the original message to determine
 * which action handler will be invoked.
 */
const createActionHandler = <T = any>(
  state: RunnerState,
  actions: Actions<T>,
  render: RenderFunction,
  middleware: PatchMiddleware[],
): ActionHandler => {
  const patcher = createPatcher(state, middleware, () => setNextRender(state, () => render(handler)));
  const handler = actionHandlerFactory(patcher, actions);
  return handler;
};

const DEFAULT_OPTIONS: RunnerOptions = {
  middleware: [],
  root: "#app",
};

/**
 * Create and start a new application runtime
 *
 * The runner function takes a model, actions mapping, view function, and an
 * optional root element selector (defaults to "#app"). It then kicks off the
 * render process, rendering the initial view onto the root element (root
 * element is replaced in the process).
 */
const runner = <T = any> (model: T, actions: Actions<T>, view: ViewFunction, options: RunnerOptions = {}) => {
  const opt = {
    ...DEFAULT_OPTIONS,
    ...options,
  };

  const state: RunnerState<T> = {
    model,
    nextRenderId: null,
    vnodes: is.str(opt.root) ? (document.querySelector(opt.root) as Element) : opt.root as Element | VNode,
  };

  // Prepare the engine
  const render = createRenderer(state, view);
  const actionHandler = createActionHandler<T>(
    state,
    actions,
    render,
    opt.middleware as PatchMiddleware[],
  );

  // Start rendering
  render(actionHandler);
};

export {
  PatchFunction,
  Scope,
  ModelPatcher,
  Actions,
  ActionHandler,
  PatchMiddleware,
  Props,
  ViewFunction,
  RunnerState,
  RunnerOptions,
  runner,
};
export default runner;
