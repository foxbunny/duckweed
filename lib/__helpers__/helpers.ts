/**
 * (c) 2017 Hajime Yamasaki Vukelic
 * All rights reserved.
 */

import {PatchFunction} from "../runner";

const pause = (delay: number = 0): Promise<void> => {
  return new Promise((res) => {
    setTimeout(res, delay);
  });
};

const modelSnapshotter = (() => {
  let modelSnapshots: Array<[any, any]> = [];
  return {
    get snapshots() {
      return modelSnapshots;
    },
    clear() {
      modelSnapshots = [];
    },
    middleware(fn: PatchFunction) {
      return (model: any) => {
        const before = model;
        const after = fn(model);
        modelSnapshots.push([before, after]);
        return after;
      };
    },
  };
})();

export {
  pause,
  modelSnapshotter,
};
