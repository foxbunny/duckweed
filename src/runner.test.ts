/**
 * (c) 2017 Hajime Yamasaki Vukelic
 * All rights reserved.
 */

import * as snabbdom from "snabbdom";
import h from "snabbdom/h";
import eventlisteners from "snabbdom/modules/eventlisteners";
import props from "snabbdom/modules/props";

import runner from "./runner";

import {modelSnapshotter, pause} from "./__helpers__/helpers";

const patch = snabbdom.init([eventlisteners, props]);

describe("runner", () => {
  beforeEach(() => modelSnapshotter.clear());

  it("Should render the initial view", () => {
    const root = document.createElement("div");
    const view = () => h("div", "Hello, root");
    runner(undefined, {}, view, {root});
    expect(root).toMatchSnapshot();
  });

  it("Should pass the model to the view as props", () => {
    const root = document.createElement("div");
    const m = {iam: "model"};
    const view = ({model}) => {
      expect(model.iam).toBe("model");
      return h("div", `Hello, ${model.iam}`);
    };
    runner(m, {}, view, {root});
    expect(root).toMatchSnapshot();
  });

  it("Should pass the action handler to view as props", () => {
    const root = document.createElement("div");
    const a = {
      foo: jest.fn((x) => x),
    };
    const view = ({act}) => {
      act("foo", "arg")();
      return h("div", `Hello, action`);
    };
    runner(undefined, a, view, {root});
    expect(a.foo).toHaveBeenCalled();
  });

  it("Should have action handler handle events", () => {
    const root = document.createElement("div");
    const a = {
      foo: jest.fn((x) => x),
    };
    const view = ({act}) => {
      return h("div", [
        h("a", {props: {href: "/foo"}, on: {click: act("foo", "bar")}}),
      ]);
    };
    runner(undefined, a, view, {root});
    const event = new Event("click");
    root.querySelector("a").dispatchEvent(event);
    expect(a.foo).toHaveBeenCalled();
    expect(a.foo.mock.calls[0][1]).toBe("bar");
    expect(a.foo.mock.calls[0][2]).toBe(event);
  });

  it("Should pass a patcher for action to mutate the model", async () => {
    const root = document.createElement("div");
    const m = 0;
    const a = {
      foo: (patcher) => {
        patcher((model) => model + 1);
      },
    };
    const view = ({act}) => {
      return h("div", [
        h("a", {props: {href: "/foo"}, on: {click: act("foo")}}),
      ]);
    };
    runner(m, a, view, {root, middleware: [modelSnapshotter.middleware]});
    const event = new Event("click");
    root.querySelector("a").dispatchEvent(event);
    await pause();
    root.querySelector("a").dispatchEvent(event);
    await pause();
    expect(modelSnapshotter.snapshots).toEqual([[0, 1], [1, 2]]);
  });

  xit("Should fail loudly if handler is passed a non-existent action", async () => {
    const root = document.createElement("div");
    const view = ({act}) => {
      return h("div", [
        h("a", {props: {href: "/foo"}, on: {click: act("foo")}}),
      ]);
    };
    runner(undefined, {}, view, {root});
    root.querySelector("a").dispatchEvent(new Event("click"));
    // FIXME: how do we test this anyway?
  });

  it("Should pass an action handler that can generate a subaction", () => {
    const root = document.createElement("div");

    // Submodule

    const suba = {
      bar: jest.fn(),
    };
    const subview = ({act}) => {
      return h("div", [
        h("a", {props: {href: "/foo"}, on: {click: act("bar")}}),
      ]);
    };

    // Main module

    const a = {
      foo: () => {
        suba.bar();
      },
    };
    const view = ({act}) => {
      return h("div", [
        subview({act: act.as("foo")}),
      ]);
    };

    runner(undefined, {}, view, {root});
    root.querySelector("a").dispatchEvent(new Event("click"));
  });
});
