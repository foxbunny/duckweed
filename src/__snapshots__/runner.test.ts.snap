// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runner Should accept a plugin function which can control the app from within the runner 1`] = `
<div>
  Pluginized
</div>
`;

exports[`runner Should pass the model to the view as props 1`] = `
<div>
  Hello, model
</div>
`;

exports[`runner Should provide a way to create scoped patchers 1`] = `
Array [
  Array [
    Object {
      "list": Object {
        "counters": Array [
          Object {
            "count": 0,
          },
          Object {
            "count": 0,
          },
          Object {
            "count": 0,
          },
          Object {
            "count": 0,
          },
        ],
        "total": 0,
      },
    },
    Object {
      "list": Object {
        "counters": Array [
          Object {
            "count": 1,
          },
          Object {
            "count": 0,
          },
          Object {
            "count": 0,
          },
          Object {
            "count": 0,
          },
        ],
        "total": 1,
      },
    },
  ],
  Array [
    Object {
      "list": Object {
        "counters": Array [
          Object {
            "count": 1,
          },
          Object {
            "count": 0,
          },
          Object {
            "count": 0,
          },
          Object {
            "count": 0,
          },
        ],
        "total": 1,
      },
    },
    Object {
      "list": Object {
        "counters": Array [
          Object {
            "count": 1,
          },
          Object {
            "count": 1,
          },
          Object {
            "count": 0,
          },
          Object {
            "count": 0,
          },
        ],
        "total": 2,
      },
    },
  ],
  Array [
    Object {
      "list": Object {
        "counters": Array [
          Object {
            "count": 1,
          },
          Object {
            "count": 1,
          },
          Object {
            "count": 0,
          },
          Object {
            "count": 0,
          },
        ],
        "total": 2,
      },
    },
    Object {
      "list": Object {
        "counters": Array [
          Object {
            "count": 1,
          },
          Object {
            "count": 1,
          },
          Object {
            "count": 1,
          },
          Object {
            "count": 0,
          },
        ],
        "total": 3,
      },
    },
  ],
]
`;

exports[`runner Should render the initial view 1`] = `
<div>
  Hello, root
</div>
`;
